experimental/ast/path.go

```diff
@@ -0,0 +1,210 @@
+	withContext
+
+	raw rawPath
+}
+
+// Absolute returns whether this path starts with a dot.
+func (p Path) Absolute() bool {
+	var abs bool
+	p.Components(func(c PathComponent) bool {
+		abs = !c.Separator().Nil()
+		return false
+	})
+	return abs
+}
+
+// AsIdent returns the single identifier that comprises this path, or
+// the nil token.
+func (p Path) AsIdent() Token {
+	var tok Token
+	var count int
+	p.Components(func(c PathComponent) bool {
+		if count > 0 {
+			tok = Token{}
+			return false
+		}
+
+		if c.Separator().Nil() {
+			tok = c.AsIdent()
+		}
+
+		count++
+		return true
+	})
+	return tok
+}
+
+// AsBuiltin returns the builtin that this path represents.
+//
+// If this path does not represent a builtin, returns [BuiltinUnknown].
+func (p Path) AsBuiltin() Builtin {
+	return BuiltinByName(p.AsIdent().Text())
+}
+
+// Span implements [Spanner] for Path.
+func (p Path) Span() Span {
+	return JoinSpans(p.raw[0].With(p), p.raw[1].With(p))
+}
+
+// Components is an [iter.Seq] that ranges over each component in this path. Specifically,
+// it yields the (nilable) dot that precedes the component, and the identifier token.
+func (p Path) Components(yield func(PathComponent) bool) {
+	if p.Nil() {
+		return
+	}
+
+	first := p.raw[0].With(p)
+	if synth := first.synthetic(); synth != nil {
+		panic("synthetic paths are not implemented yet")
+	}
+
+	cursor := Cursor{
+		withContext: p.withContext,
+		start:       p.raw[0],
+		end:         p.raw[1] + 1, // Remember, Cursor.end is exclusive!
+	}
+
+	var sep Token
+	var broken bool
+	cursor.Iter(func(tok Token) bool {
+		if tok.Text() == "." || tok.Text() == "/" {
+			if !sep.Nil() {
+				// Uh-oh, empty path component!
+				if !yield(PathComponent{sep, Token{}}) {
+					broken = true
+					return false
+				}
+			}
+			sep = tok
+			return true
+		}
+
+		if !yield(PathComponent{sep, tok}) {
+			broken = true
+			return false
+		}
+		sep = Token{}
+		return true
+	})
+	if !broken && !sep.Nil() {
+		yield(PathComponent{sep, Token{}})
+	}
+}
+
+// PathComponent is a piece of a path. This is either an identifier or a nested path
+// (for an extension name).
+type PathComponent struct {
+	separator, name Token
+}
+
+// Separator is the token that separates this component fromm the previous one, if
+// any. This may be a dot or a slash.
+func (p PathComponent) Separator() Token {
+	return p.separator
+}
+
+// Name is the token that represents this component's name. THis is either an
+// identifier or a (...) token containing a path.
+func (p PathComponent) Name() Token {
+	return p.name
+}
+
+// Returns whether this is an empty path component. Such components are not allowed
+// in the grammar but may occur in invalid inputs nonetheless.
+func (p PathComponent) IsEmpty() bool {
+	return p.Name().Nil()
+}
+
+// IsExtension returns whether this path component is an extension component, i.e.
+// (a.b.c).
+func (p PathComponent) IsExtension() bool {
+	return p.Name().Kind() == TokenPunct
```

Comment:
I think this would be more clear to think about this case being whether the name token is non-leaf. So leaf tokens are the path element and non-leaf tokens contain a path for the extension. 
```suggestion
	return !p.Name().IsLeaf()
```
But I think that also means that this returns true for the URL "field" of an Any message literal that uses the Any expanded text syntax, right? That's probably at least worth mentioning in the doc? Or maybe even worth renaming this to some more-general concept that can encompass both Any URLs and extensions?

Reply:
Well. If you have an Any path like `foo.com/my.Type`, as path components, it will look like `[(_, foo), (., com), (/, my), (., Type)]`. So this is not a question that's particularly suitable to answer with this abstraction. I expect that an API for handling this case will appear as we figure out how the middle-end of the compiler will look.

I will add a comment noting this though.